ssm/lds.py

```diff
@@ -609,52 +659,129 @@ def hessian_neg_expected_log_joint(x, Ez, Ezzp1, scale=1):
+                    return self._laplace_neg_expected_log_joint(data,
+                                                                input,
+                                                                mask,
+                                                                tag,
+                                                                x,
+                                                                Ez,
+                                                                Ezzp1,
+                                                                scale=scale)
+
                 x = lbfgs(_objective, x0, num_iters=continuous_maxiter,
                           tol=continuous_tolerance)
 
             # Evaluate the Hessian at the mode
-            assert np.all(np.isfinite(obj(x)))
-            J_diag, J_lower_diag = hessian_neg_expected_log_joint(x, Ez, Ezzp1)
-
-            # Compute the Hessian vector product h = J * x = -H * x
-            # We can do this without instantiating the full matrix
-            h = symm_block_tridiag_matmul(J_diag, J_lower_diag, x)
+            assert np.all(np.isfinite(obj(x))) 
+
+            J_ini, J_dyn_11, J_dyn_21, J_dyn_22, J_obs = self.\
+                _laplace_neg_hessian_params(data, input, mask, tag, x, Ez, Ezzp1)
+            h_ini, h_dyn_1, h_dyn_2, h_obs = \
+                self._laplace_neg_hessian_params_to_hs(x, J_ini, J_dyn_11,
+                                              J_dyn_21, J_dyn_22, J_obs)
+
+            prms["J_ini"] = J_ini
+            prms["J_dyn_11"] = J_dyn_11
+            prms["J_dyn_21"] = J_dyn_21
+            prms["J_dyn_22"] = J_dyn_22
+            prms["J_obs"] = J_obs
+            
+            prms["h_ini"] = h_ini
+            prms["h_dyn_1"] = h_dyn_1
+            prms["h_dyn_2"] = h_dyn_2
+            prms["h_obs"] = h_obs
+
+    def _fit_laplace_em_params_update_exact_mstep(
+            self, expectations, continuous_samples,
+            datas, inputs, masks, tags,
+            emission_optimizer, emission_optimizer_maxiter,
+            alpha, continuous_expectations=None,):
+
+        # For now we can only do the exact update with linear-Gaussian dynamics
+        # and lags == 1. This check should pass as long as the dynamics is a
+        # subclass of AutoRegressiveObservations (e.g
+        # AutoRegressiveObservationsNoInput)
+        assert isinstance(self.dynamics, obs.AutoRegressiveObservations), \
+            "We can only do an exact m-step with Linear-Gaussian Dynamics."
+
+        # 3. Update the model parameters exactly
+        xmasks = [np.ones_like(x, dtype=bool) for x in continuous_samples]
+        for distn in [self.init_state_distn, self.transitions]:
+            if distn.params == tuple(): continue
+            distn.m_step(expectations, continuous_samples, inputs, xmasks,
+                         tags)
+        self.dynamics.m_step(expectations, continuous_samples, inputs, xmasks,
```

Comment:
Just noticed this... shouldn't we be using `continuous_expectations` here?

Reply:
yeah you're right. That's a pretty big oversight on my part since it means we haven't been using the exact m-step codepath when testing things. I will fix this and make sure things still seem to be working.